mava/systems/ppo/sebulba/rec_ippo.py

```diff
@@ -0,0 +1,866 @@
+    ) -> Tuple[RNNLearnerState, Metrics]:
+        """Learner function.
+
+        This function represents the learner, it updates the network parameters
+        by iteratively applying the `_update_step` function for a fixed number of
+        updates. The `_update_step` function is vectorized over a batch of inputs.
+
+        Args:
+            learner_state (NamedTuple):
+                - params (Params): The initial model parameters.
+                - opt_states (OptStates): The initial optimizer state.
+                - key (chex.PRNGKey): The random number generator state.
+                - env_state (LogEnvState): The environment state.
+                - timesteps (TimeStep): The last timestep of the rollout.
+                - dones (bool): Whether the initial timestep was a terminal state.
+                - hstateS (HiddenStates): The initial hidden states of the RNN.
+        """
+        # This function is shard mapped on the batch axis, but `_update_step` needs
+        # the first axis to be time
+        traj_batch = tree.map(switch_leading_axes, traj_batch)
+        learner_state, loss_info = _update_step(learner_state, traj_batch)
+
+        return learner_state, loss_info
+
+    return learner_fn
+
+
+def learner_thread(
+    learn_fn: SebulbaLearnerFn[RNNLearnerState, RNNPPOTransition],
+    learner_state: RNNLearnerState,
+    config: DictConfig,
+    eval_queue: Queue,
+    pipeline: Pipeline,
+    params_sources: Sequence[ParamsSource],
+) -> None:
+    for _ in range(config.arch.num_evaluation):
+        # Create the lists to store metrics and timings for this learning iteration.
+        metrics: List[Tuple[Dict, Dict]] = []
+        rollout_times_array: List[Dict] = []
+        learn_times: Dict[str, List[float]] = defaultdict(list)
+
+        with RecordTimeTo(learn_times["learner_time_per_eval"]):
+            for _ in range(config.system.num_updates_per_eval):
+                # Get the trajectory batch from the pipeline
+                # This is blocking so it will wait until the pipeline has data.
+                with RecordTimeTo(learn_times["rollout_get_time"]):
+                    traj_batch, rollout_time, ep_metrics, (timestep, hstates) = pipeline.get(  # type: ignore
+                        block=True
+                    )
+
+                # Replace the timestep in the learner state with the latest timestep
+                # This means the learner has access to the entire trajectory as well as
+                # an additional timestep which it can use to bootstrap.
+                learner_state = learner_state._replace(timestep=timestep)
+                learner_state = learner_state._replace(
+                    dones=timestep.last()
+                    .repeat(config.system.num_agents)
+                    .reshape(config.arch.num_envs, -1)
+                )
+                learner_state = learner_state._replace(hstates=hstates)  # type: ignore
```

Comment:
type ignore might be in the wrong place, but I think this should work
```suggestion
                learner_state = learner_state._replace(
                    timestep=timestep,
                    dones=timestep.last()
                    .repeat(config.system.num_agents)
                    .reshape(config.arch.num_envs, -1),
                    hstates=hstates
                )  # type: ignore
```